src/main/java/edu/kit/compiler/optimizations/inlining/Inliner.java

```diff
@@ -0,0 +1,220 @@
+        copyVisitor.replacePreds();
+
+        Node[] phis = copyVisitor.createPhis();
+        for (var edge: BackEdges.getOuts(call)) {
+            Proj proj = (Proj) edge.node;
+            if (proj.getMode().equals(Mode.getM())) {
+                Graph.exchange(proj, phis[0]);
+            } else if (proj.getMode().equals(Mode.getT())) {
+                for (var retEdge: BackEdges.getOuts(proj)) {
+                    // handle weird endless loop edge case
+                    Node newNode = phis.length > 1 ? phis[1]
+                            : graph.newConst(0, retEdge.node.getMode());
+                    Graph.exchange(retEdge.node, newNode);
+                }
+            }
+        }
+
+        splitBlock(call, copyVisitor.getEndBlock(), jmp);
+        BackEdges.disable(callee);
+    }
+
+    private static void splitBlock(Call call, Node endBlock, Node entryJmp) {
+        // mark all nodes that remain in the current block
+        Set<Integer> callPreds = new HashSet<>();
+        callPreds.add(entryJmp.getNr());
+        markPreds(callPreds, call, call.getBlock());
+
+        // move the nodes to the new block
+        for (var edge: BackEdges.getOuts(call.getBlock())) {
+            Node node = edge.node;
+            if (node.getBlock().equals(call.getBlock()) && !callPreds.contains(node.getNr())
+                    && node.getOpCode() != ir_opcode.iro_Phi
+                    && !mustBeInStartBlock(node)) {
+                if (node.getOpCode() != ir_opcode.iro_Proj) {
+                    node.setBlock(endBlock);
+                } else if (!callPreds.contains(getNonProjPred(node).getNr())) {
+                    // projections must be in same block as predecessor
+                    node.setBlock(endBlock);
+                }
+            }
+        }
+    }
+
+    private static void markPreds(Set<Integer> marked, Node current, Node block) {
+        marked.add(current.getNr());
+        for (int i = 0; i < current.getPredCount(); i++) {
+            Node pred = current.getPred(i);
+            if (!marked.contains(pred.getNr()) && pred.getBlock().equals(block)) {
+                markPreds(marked, pred, block);
+            }
+        }
+    }
+
+    /**
+     * Find the first predecessor of the projection that is not
+     * itself a projection.
+     */
+    private static Node getNonProjPred(Node node) {
+        assert node.getOpCode() == ir_opcode.iro_Proj;
+        Node pred = ((Proj) node).getPred();
+        if (pred.getOpCode() == ir_opcode.iro_Proj) {
+            return getNonProjPred(pred);
+        } else {
+            return pred;
+        }
+    }
+
+    private static boolean mustBeInStartBlock(Node node) {
+        return node.getOpCode() == ir_opcode.iro_Const
+                || node.getOpCode() == ir_opcode.iro_Address
+                || node.getOpCode() == ir_opcode.iro_NoMem;
+    }
+
+    @RequiredArgsConstructor
+    private static class CopyNodes extends NodeVisitor.Default {
```

Comment:
I think you may need a special case for `NoMem`. Like with the initial memory, there must only be a single such node per graph.